### DISCUSSION TITLE
Is there a way to add image height and width in theme automagically

### DISCUSSION THREAD

**Discussion by aledeg** [getnikola/nikola#3547] [Q&A] [answered]
When checking for accessibility, one of the things that bothers me its the warning about image sizes. I search how to do that without including it manually on all included images. I was wondering if there is something on theme level I can do to address that.

**Comment by Kwpolska** [comment, marked as answer]
Nikola can’t do that automagically, and that probably wouldn’t be done on the theme level. It could be done by post compilers, I don’t think reST has an option for this. You could try extending your input format of choice with that feature, or perhaps a filter — but note that reading the images to get their sizes might increase build time a lot (especially if you try to do that with remote images.)

**Reply by aledeg** [reply]
Actually, I know the image size but not the orientation. I am trying to add that on my thumbnails.
I guess I will have to do it by hand.
Thank you!

**Reply by Kwpolska** [reply]
Figuring out the orientation would still require reading the images.

**Reply by felixfontein** [reply]
For most sane images formats, it suffices to read the header / metadata to get that information. For example, Pillow's [`Image.open()`](https://pillow.readthedocs.io/en/latest/reference/Image.html#functions) only reads metadata (until you explicitly tell the resulting object to load the pixel data), and you can use the `width` and `height` properties of that object without loading the pixel data. That's pretty quick even if you have quite a few images.

For example, I ran
```.py
import os
from PIL import Image

data = []
Image.MAX_IMAGE_PIXELS = None
for p, _, fns in os.walk('.'):
    for fn in fns:
        if not fn.endswith(('.png', '.jpg', '.jpeg', '.JPG', '.JPEG')):
            continue
        fn = os.path.join(p, fn)
        try:
            with Image.open(fn) as i:
                data.append((fn, i.width, i.height))
        except Exception as e:
            print('{0}: {1}'.format(fn, e))
print(len(data))
```
in a folder with ~5500 files, it took 0.56 seconds.

When I ran it in a larger photo collection (~750 GB, ~140'000 files), it took 5.5 minutes. So for a certain amount of images, caching makes sense :) But still, it is *a lot* faster than actually reading all that image data :)
